src/main/scala/firrtl/passes/NewLowerTypes.scala

```diff
@@ -0,0 +1,439 @@
+// See LICENSE for license details.
+
+package firrtl.passes
+
+import firrtl.analyses.InstanceKeyGraph
+import firrtl.annotations.{CircuitTarget, MemoryInitAnnotation, MemoryRandomInitAnnotation, ModuleTarget, ReferenceTarget}
+import firrtl.{CircuitForm, CircuitState, InstanceKind, Kind, MemKind, PortKind, RenameMap, SymbolTable, Transform, UnknownForm, Utils}
+import firrtl.ir._
+import firrtl.options.Dependency
+import firrtl.stage.TransformManager.TransformDependency
+
+import scala.collection.mutable
+
+/** Flattens Bundles and Vecs.
+  * - Some implicit bundle types remain, but with a limited depth:
+  *   - the type of a memory is still a bundle with depth 2 (mem -> port -> field), see [[MemPortUtils.memType]]
+  *   - the type of a module instance is still a bundle with depth 1 (instance -> port)
+  */
+object NewLowerTypes extends Transform {
+  override def inputForm: CircuitForm = UnknownForm
+  override def outputForm: CircuitForm = UnknownForm
+
+  override def prerequisites: Seq[TransformDependency] = Seq(
+    Dependency(RemoveAccesses), // we require all SubAccess nodes to have been removed
+    Dependency(CheckTypes),     // we require all types to be correct
+    Dependency(InferTypes),     // we require instance types to be resolved (i.e., DefInstance.tpe != UnknownType)
+    Dependency(ExpandConnects)  // we require all PartialConnect nodes to have been expanded
```

Comment:
So does this work prior to expand whens?

Reply:
I haven't tested it but I haven't been able to find a fundamental reason why it shouldn't. As long as all references are lowered to ground types (ExpandConnects) and there are no SubAccess nodes, I think we should be fine.